### DISCUSSION TITLE
ZUGFeRDVisualizer toFOP and toPDF overload or refactoring idea

### DISCUSSION THREAD

**Discussion by taskinkemal** [ZUGFeRD/mustangproject#537] [Ideas]
Hi all, at the moment toPDF and toFOP methods work inherently with files. In my local copy I'm working on the idea of working directly with streams and with string input. Something like this:

byte[] toPDFFromString(String xmlContent); // this uses toFOPFromString internally.
String toFOPFromString(String xmlContent);

In my copy I cloned the original methods and adjusted them to my needs. But for sure I can do a refactoring and reuse from the original methods to prevent code duplication.

Does it make sense, or is it worth it to move forward and contribute to the repo with this idea?

**Comment by jstaerk** [comment]
Hello,
A PR would be highly appreciated, but keep in mind that there are three places where you need to be able to pass a stream: in both input xslts where the intermediate format is build (input being cii or ubl), in the XSLT where the final fop is build and within apache fop where the pdf is converted from the fop.
kind regards
Jochen

**Reply by taskinkemal** [reply]
Thank you,

True, one should pass the input stream to all three.

I created a very small PR to make toFOP method reusable. This removes the dependency to the file system and enables future improvements in this direction. I wanted to it with small steps.

I would appreciate if you can kindly take a look and evaluate. Thanks.

**Reply by jstaerk** [reply]
I'll have a look

**Reply by taskinkemal** [reply]
Hi, please take a look at this PR:
https://github.com/ZUGFeRD/mustangproject/pull/597
This was what I had in mind: An alternative to toPDF, which doesn't depend on the file system. XML string in, byte array out.
If it makes sense, we can continue from the PR. Thank you for your attention.

**Reply by jstaerk** [reply]
> Hi, please take a look at this PR: #597 This was what I had in mind: An alternative to toPDF, which doesn't depend on the file system. XML string in, byte array out. If it makes sense, we can continue from the PR. Thank you for your attention.

looks good just seems to have a small conflict

**Reply by taskinkemal** [reply]
I'll take a look now.

**Reply by taskinkemal** [reply]
Hi, the PR with the resolved conflicts is here:
https://github.com/ZUGFeRD/mustangproject/pull/649
Thank you.
